db/compaction_job.cc

```diff
@@ -1212,10 +1212,11 @@ Status CompactionJob::FinishCompactionOutputFile(
     for (; it->Valid(); it->Next()) {
       auto tombstone = it->Tombstone();
       if (upper_bound != nullptr &&
-          ucmp->Compare(*upper_bound, tombstone.start_key_) <= 0) {
-        // Tombstones starting at upper_bound or later only need to be included
-        // in the next table. Break because subsequent tombstones will start
-        // even later.
+          ucmp->Compare(*upper_bound, tombstone.start_key_) < 0) {
```

Comment:
@abhimadan @ajkr This change broke one of the tests in CockroachDB. The test sets up a scenario where there are 4 sstables:

```
0: "a000000000" - "c000010000"
6: "a000000000" - "a000009999"
6: "b000000000" - "b000009999"
6: "c000000000" - "c000009999"
```

The L0 sstable has a single key at `a000000000` and a range deletion from `[c000000000,c000010000)`. The test then performs a manual compaction from `[c000000000,c000010000)` (i.e. the same keys as the range deletion). The first step in the manual compaction compacts the L0 sstable to L5 which results in two sstables at L5:

```
#12: a000000000#1,1 c000000000#72057594037927935,15
#13: c000000000#3,0 c000010000#72057594037927935,15
```

This is where the problem occurs for the test. Prior to this line of code changing, the compaction generated L5 sstables that looked like:

```
#12: a000000000#1,1 a000000000#1,1
#13: c000000000#3,0 c000010000#72057594037927935,15
```

It looks like the range tombstone is being included in sstable `#12` unnecessarily. I see that this change was made in order to fix the problem tested by `KeyAtOverlappingEndpointReappears`, though it looks to be pessimistic. In my test, `upper_bound` is being set to the min-key in the following table.

Reply:
@petermattis - thanks for pointing this out. Just to confirm, the problem here is performance (we store an extra range tombstone and unnecessarily extend an SST file), not correctness, right?

Understood what you meant about comparing against next file's start key being pessimistic. Let's see if there's a way to compare against the current file's end key instead.

Reply:
On second thought, then we'd need additional logic to handle tombstones fully in the gap between files. Will think a bit more.

Reply:
Yes, the problem is performance, not correctness. In my test, when `#12` is compacted from L5 to L6, it will compact against 2 sstables in L6 instead of 1. I think we can generate scenarios which produce arbitrarily large compactions due to this.

Reply:
OK, we can probably revert this change, and fix `KeyAtOverlappingEndpointReappears` by having an additional condition that tombstones we skip should start strictly after the file's last point key.

Reply:
proposed fix: #4592. BTW, if your test exercises RocksDB APIs directly, I'd be interested in copying it :P.